extensions/socket/internal.m

```diff
@@ -0,0 +1,362 @@
+    }
+
+    NSError *err;
+    if (![asyncSocket connectToHost:theHost onPort:[thePort unsignedShortValue] error:&err]) {
+        [skin logError:[NSString stringWithFormat:@"Unable to connect: %@", err]];
+    }
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket.read(delimiter) -> self
+/// Method
+/// Read data from the socket. Data is passed to the callback function
+///
+/// Parameters:
+///  * delimiter - Either a number of bytes to read, or a string delimiter such as `\n` or `\r\n`. Data is read up to and including the delimiter
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_read(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TSTRING|LS_TNUMBER, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
+    if (!asyncSocket.callback) {
+        [skin logWarn:@"No callback! Defaulting to print()"];
+    }
+
+    switch (lua_type(L, 2)) {
+        case LUA_TNUMBER: {
+            NSNumber *bytesToRead = [skin toNSObjectAtIndex:2];
+            NSUInteger bytes = [bytesToRead unsignedIntegerValue];
+            [asyncSocket readDataToLength:bytes withTimeout:-1 tag:-1];
+        } break;
+
+        case LUA_TSTRING: {
+            NSString *separatorString = [skin toNSObjectAtIndex:2];
+            NSData *separator = [separatorString dataUsingEncoding:NSUTF8StringEncoding];
+            [asyncSocket readDataToData:separator withTimeout:-1 tag:-1];
+        } break;
+
+        default:
+            break;
+    }
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket.write(message) -> self
+/// Method
+/// Write data to the socket
+///
+/// Parameters:
+///  * message - A string containing data to be sent on the socket
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_write(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TSTRING, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+    NSString *message = [skin toNSObjectAtIndex:2];
+
+    [asyncSocket writeData:[message dataUsingEncoding:NSUTF8StringEncoding ] withTimeout:-1 tag:-1];
+
+    [skin pushNSObject:asyncSocket];
+    return 1;
+}
+
+/// hs.socket:setCallback([fn]) -> self
+/// Method
+/// Sets the callback for the socket. Required for working with read data
+///
+/// Parameters:
+///  * fn - An optional callback function with single parameter containing data read from the socket. A `nil` argument or nothing clears the callback, defaulting to `print`
+///
+/// Returns:
+///  * The `hs.socket` object
+///
+static int socket_setCallback(lua_State *L) {
+    LuaSkin *skin = [LuaSkin shared];
+    [skin checkArgs:LS_TUSERDATA, USERDATA_TAG, LS_TFUNCTION|LS_TNIL|LS_TOPTIONAL, LS_TBREAK];
+
+    HSAsyncSocket*  asyncSocket = [skin luaObjectAtIndex:1 toClass:"HSAsyncSocket"];
+
```

Comment:
Before doing anything with the LUA_TFUNCTION, this should do asyncSocket.callback = [skin luaUnref:refTable asyncSocket.callback] - it's a no-op if the callback is LUA_NOREF, but if a value has been stored, it is released from the reference table before we add another one